mep/books/migrations/0022_generate_work_slugs.py

```diff
@@ -0,0 +1,72 @@
+# Generated by Django 2.2.11 on 2020-03-31 19:24
+
+from collections import defaultdict
+
+from django.db import migrations, models
+
+from mep.books.utils import nonstop_words, work_slug
+
+
+def generate_slugs(apps, schema_editor):
+    '''Generate uniuqe slugs for works'''
+    Work = apps.get_model('books', 'Work')
+
+    # first pass: generate slugs
+    works = Work.objects.all()
+    for work in works:
+        work.slug = work_slug(work)
+        work.save()
+
+    # second pass: make slugs unique
+    # - get a list of all slugs that occur more than once
+    duplicate_slugs = Work.objects.values('slug') \
+        .annotate(slug_count=models.Count('slug')) \
+        .filter(slug_count__gt=1).values_list('slug', flat=True)
+
+    # - find all works by slug and uniquify
+    for dupe_slug in duplicate_slugs:
+        # counter for slugs that require numbers to differentiate
+        group_slugs = defaultdict(int)
+        # - get all the works with that slug, ordered by title
+        for work in Work.objects.filter(slug=dupe_slug).order_by('title') \
+                                                       .distinct():
+
+            nonstop_title_words = nonstop_words(work.title)
+            # if title has more than three words, use the 4th for uniqueness
+            if len(nonstop_title_words) > 3:
+                work.slug = work_slug(work, max_words=4)
+
+                # if 4-word title slug is still not unique, try 5
+                if work.slug in group_slugs and len(nonstop_title_words) > 4:
+                    work.slug = work_slug(work, max_words=5)
+
+            # count this slug
+            group_slugs[work.slug] += 1
+
+            if group_slugs[work.slug] > 1:
+                # slug is already in use; add a number to differentiate
+                if group_slugs[work.slug] > 1:
+                    work.slug = '%s-%d' % (work.slug, group_slugs[work.slug])
+
```

Comment:
to clarify I understand this:
- for all works in a set that have duplicate slugs:
- try extending the title portion by one word
- if that doesn't produce uniqueness, try again
- if that fails, just add a number to produce a unique string

this is an interesting approach but it makes sense!